django/contrib/admin/views/autocomplete.py

```diff
@@ -0,0 +1,70 @@
+from django.http import Http404, JsonResponse
+from django.views.generic.list import BaseListView
+
+
+class AutocompleteJsonView(BaseListView):
+    """
+    View that handles requests from `.AutocompleteWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginate_by = 20
+    model_admin = None
+
+    def has_perm(self, request, obj=None):
+        """Check if user has permission to access the related model."""
+        return self.model_admin.has_change_permission(request, obj=obj)
+
+    def get_paginator(self, queryset, per_page, orphans=0,
+                      allow_empty_first_page=True, **kwargs):
+        return self.model_admin.get_paginator(self.request, queryset, per_page, orphans, allow_empty_first_page)
+
+    def get(self, request, *args, **kwargs):
+        """
+        Return JSON search result based on a field identifier and a search term.
+
+        Returns:
+             (django.http.JsonResponse)::
+                {
+                    results: [
+                        {
+                            id: "123"
+                            text: "foo",
+                        }
+                    ],
+                    pagination: {
+                        more: true
+                    }
+                }
+        """
+        self.term = request.GET.get('term', '')
+        self.paginator_class = self.model_admin.paginator
+
+        if not self.has_perm(request):
+            raise Http404
```

Comment:
Add a message please.

Reply:
We did that on purpose. If we raise a message here, I would suggest to jump back to a 403, permission denied. What do you think?

Reply:
The message should only be visible when DEBUG=True so I don't think there's a concern about information leakage if you raise 404 with a message.

Reply:
I already changed it back to a `403 Forbidden`, since the permission is denied.
I can change it to anything else, but I believe this to be more correct and intuitive.